pkg/content/gunzip.go

```diff
@@ -0,0 +1,63 @@
+package content
+
+import (
+	"compress/gzip"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/containerd/containerd/content"
+)
+
+// NewGunzipWriter wrap a writer with a gunzip, so that the stream is gunzipped
+func NewGunzipWriter(writer content.Writer, blocksize int) content.Writer {
+	if blocksize == 0 {
+		blocksize = DefaultBlocksize
+	}
+	return NewPassthroughWriter(writer, func(r io.Reader, w io.Writer, done chan<- error) {
+		gr, err := gzip.NewReader(r)
+		if err != nil {
+			done <- fmt.Errorf("error creating gzip reader: %v", err)
+			return
+		}
+		// write out the uncompressed data
+		b := make([]byte, blocksize, blocksize)
+		for {
+			var n int
+			n, err = gr.Read(b)
+			if err != nil && err != io.EOF {
+				err = fmt.Errorf("GunzipWriter data read error: %v\n", err)
+				break
+			}
+			l := n
+			if n > len(b) {
+				l = len(b)
+			}
+			if _, err2 := w.Write(b[:l]); err2 != nil {
+				err = fmt.Errorf("GunzipWriter: error writing to underlying writer: %v", err2)
+				break
+			}
+			if err == io.EOF {
+				// clear the error
+				err = nil
+				break
+			}
+		}
+		gr.Close()
+		done <- err
+	})
+}
+
+// checkCompression check if the mediatype uses gzip compression or tar
+func checkCompression(mediaType string) (gzip, tar bool) {
+	mt := mediaType
+	gzipSuffix := "+gzip"
+	if strings.HasSuffix(mt, gzipSuffix) {
+		mt = mt[:len(mt)-len(gzipSuffix)]
+		gzip = true
+	}
+	if strings.HasSuffix(mt, ".tar") {
+		tar = true
+	}
+	return
+}
```

Comment:
is this used anywhere?

Reply:
Ah shoot. No, it isn't (for now). I have a `DecompressStore` that wraps any of the other stores (`FileStore` or `MemoryStore`, etc.) that uses the `GunzipWriter` and `UntarWriter`; it uses the above to determine what to wrap.

I will take it out so this is clean to merge in, and then we can discuss if we want to use add that store and how.

Reply:
Done. Back to you @jdolitsky